### PR TITLE
DAOS-11043 umem: Blob backed Memory(BMEM) interface for umem

### PR DESCRIPTION
**Pull request by sherintg**
This commit contains a new allocator which is a modified version of pmemobj. The new allocator operates on volatile memory and provides transaction functionality similar to that of pmemobj.

The BMEM interfaces internally use the new allocator functionalties to implement the required backend for hosting metadata on volatile memory backed by blob.

Note: The current commit default to BMEM instead of PMDK. One can override this by setting DAOS_USE_PMDK=1 in the server config file.

Allow-unstable-test: true

Signed-off-by: sherintg <sherin-t.george@hpe.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it
      should be force landed and forced-landing tag is set.
* [ ] Commit messages follows the guidelines outlined
      [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code.
* [ ] Any appropriate watchers have had a chance to review the PR.
* [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested
      changes.
* [ ] Githooks were used. If not, there is a sufficient reason to move forward.  Check copyrights
      if githooks were not used.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future
      PRs.
* [ ] All builds have passed.  Check non-required builds to ensure they don't indicate any problem
      such as a compiler warning on extraneous platforms.
* [ ] No new NLT or valgrind warnings.  Check the classic view. This step should only matter if
      the build requires force landing.
* [ ] Ensure sufficent testing is done.   Check feature pragmas and test tags. Check that tests
      skipped for the ticket are run and now pass with the changes.
* [ ] Quick-build or Quick-functional is not used.
* [ ] Pay attention to PRs that may affect compatibility between versions and ensure it has been
      addressed.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If
      it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Check the commit message when landing. Check the standard
      [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit
      it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by jolivier23 — CHANGES_REQUESTED** [review]
Looks like you hit the naming convention problem with your branch.  The test looks for a / so if you just change your branch from sherintg_DAOS11043 to sherintg/DAOS11043, that will fix it.   Alternatively, we could fix the check so it accepts an _ as well.  But regardless, testing, will not run without that change.

File: src/common/mem.c [hunk 1]
```diff
@@ -65,27 +76,38 @@ struct umem_pool *
...
+				D_ERROR("Enable SCM usage statistics failed. "DF_RC"\n",
+					DP_RC(rc));
+				rc = umem_tx_errno(rc);
+				pmemobj_close(pop);
+				pop = NULL;
+			}
+		}
+		return ((struct umem_pool *)pop);
+	}
+	dav_obj_t *pop;
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/mem.c:102]: Addressed this in mem.c. The files in dav directory are not always strictly following the above convention. The number of files to be modified  high. Will take it up as part of further cleanup.

File: src/common/dav/memops.c [hunk 2]
```diff
@@ -0,0 +1,691 @@
...
+static int
+operation_log_transient_init(struct operation_log *log)
+{
+	log->capacity = ULOG_BASE_SIZE;
+	log->offset = 0;
+
+	struct ulog *src = Zalloc(sizeof(struct ulog) +
+		ULOG_BASE_SIZE);
+	if (src == NULL) {
+		D_CRIT("Zalloc!\n");
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav/memops.c:97]: Changed alloc.h to reflect the same. Also removed alloc.c from the build. Now Malloc, Zalloc, Realloc and Free point to the DAOS macros.

**Comment by sherintg** [comment]
Will be creating a new PR with the right branch name shortly.

**Pull request closed without merging**